File: capstone_project/lib/services/tts_service.dart
import 'dart:async';
import 'dart:collection';
import 'package:flutter/foundation.dart';
import 'package:flutter_tts/flutter_tts.dart';
import 'package:http/http.dart' as http;
import '../core/constants/app_constants.dart';

class _QueueItem {
  final String key;
  final int priority;
  final DateTime ts;
  _QueueItem(this.key, this.priority) : ts = DateTime.now();
}

class TtsService {
  final FlutterTts _tts = FlutterTts();
  String _lang = 'en';
  bool _speaking = false;
  final List<_QueueItem> _queue = [];
  final Map<String, DateTime> _lastSpoken = {};
  DateTime _globalLast = DateTime(2000);

  // Umuganda TTS base URL — from .env
  // [PLACEHOLDER: set UMUGANDA_HOST and UMUGANDA_PORT in .env]
  final String _umugandaUrl = 'http://127.0.0.1:5002/api/tts';

  Future<void> initialize() async {
    await _tts.setLanguage('en-US');
    await _tts.setSpeechRate(0.85);
    await _tts.setPitch(1.0);
    await _tts.setVolume(1.0);
    _tts.setCompletionHandler(() {
      _speaking = false;
      _processQueue();
    });
    _tts.setErrorHandler((msg) {
      _speaking = false;
      _processQueue();
    });
  }

  void setLanguage(String lang) {
    _lang = lang;
    stop();
  }

  void enqueue(String promptKey, int priority) {
    final last = _lastSpoken[promptKey];
    if (last != null &&
        DateTime.now().difference(last).inMilliseconds <
            AppConstants.feedbackCooldownMs) return;

    // Critical prompts displace lower-priority items
    if (priority == AppConstants.priorityCritical) {
      _queue.removeWhere((i) => i.priority > AppConstants.priorityCritical);
    }

    _queue.add(_QueueItem(promptKey, priority));
    _queue.sort((a, b) =>
        a.priority != b.priority ? a.priority - b.priority : a.ts.compareTo(b.ts));

    if (!_speaking) _processQueue();
  }

  void stop() {
    _tts.stop();
    _queue.clear();
    _speaking = false;
  }

  void _processQueue() {
    if (_speaking || _queue.isEmpty) return;
    final elapsed = DateTime.now().difference(_globalLast).inMilliseconds;
    if (elapsed < AppConstants.feedbackMinIntervalMs) {
      Future.delayed(
        Duration(milliseconds: AppConstants.feedbackMinIntervalMs - elapsed),
        _processQueue,
      );
      return;
    }
    final next = _queue.removeAt(0);
    _speak(next.key);
  }

  Future<void> _speak(String key) async {
    final text = _resolveText(key);
    if (text == null) { _processQueue(); return; }
    _speaking = true;
    _lastSpoken[key] = DateTime.now();
    _globalLast = DateTime.now();

    // Try Umuganda for Kinyarwanda
    if (_lang == 'rw') {
      final ok = await _speakUmuganda(text);
      if (ok) return;
    }

    // Fallback to flutter_tts
    await _tts.setLanguage(_lang == 'rw' ? 'rw-RW' : 'en-US');
    await _tts.speak(text);
  }

  Future<bool> _speakUmuganda(String text) async {
    try {
      final resp = await http.post(
        Uri.parse(_umugandaUrl),
        headers: {'Content-Type': 'application/json'},
        body: '{"text":"$text","speaker_id":0}',
      ).timeout(const Duration(seconds: 3));
      if (resp.statusCode != 200) return false;
      // [PLACEHOLDER: play resp.bodyBytes via audioplayer]
      // For now falls through to flutter_tts
      return false;
    } catch (_) {
      return false; // Server not running — use flutter_tts
    }
  }

  String? _resolveText(String key) {
    final map = _lang == 'rw' ? AppConstants.promptsRw : AppConstants.promptsEn;
    return map[key] ?? AppConstants.promptsEn[key];
  }
}

File: capstone_project/web/src/services/ttsService.js
// web/src/services/ttsService.js
// Priority-queued TTS using Web Speech API (EN + RW fallback)
// and optional Umuganda HTTP TTS server for Kinyarwanda.
// Mirrors the behaviour of lib/services/tts_service.dart.

import { CONST, PROMPTS_EN, PROMPTS_RW } from '../utils/constants.js';

class QueueItem {
  constructor(key, priority) {
    this.key = key;
    this.priority = priority;
    this.ts = Date.now();
  }
}

export class TTSService {
  constructor() {
    this._lang = 'en';
    this._queue = [];
    this._speaking = false;
    this._lastSpoken = {};   // { key: timestamp }
    this._globalLast = 0;
    this._synth = window.speechSynthesis ?? null;
    this._umugandaUrl = 'http://127.0.0.1:5002/api/tts';
    this._audioCtx = null;   // Created on first Umuganda response
  }

  setLang(lang) {
    this._lang = lang;
    this.stop();
  }

  /**
   * Add a feedback cue to the priority queue.
   * Lower priority number = higher urgency (mirrors Dart constants).
   */
  enqueue(key, priority) {
    const last = this._lastSpoken[key] ?? 0;
    if (Date.now() - last < CONST.FEEDBACK_COOLDOWN_MS) return;

    // Critical prompts flush lower-priority items
    if (priority === CONST.PRIORITY_CRITICAL) {
      this._queue = this._queue.filter(i => i.priority <= CONST.PRIORITY_CRITICAL);
    }

    this._queue.push(new QueueItem(key, priority));
    // Sort: ascending priority, then FIFO within same priority
    this._queue.sort((a, b) =>
      a.priority !== b.priority ? a.priority - b.priority : a.ts - b.ts
    );

    if (!this._speaking) this._processQueue();
  }

  stop() {
    this._synth?.cancel();
    this._queue = [];
    this._speaking = false;
  }

  _processQueue() {
    if (this._speaking || this._queue.length === 0) return;
    const elapsed = Date.now() - this._globalLast;
    if (elapsed < CONST.FEEDBACK_MIN_INTERVAL_MS) {
      setTimeout(() => this._processQueue(), CONST.FEEDBACK_MIN_INTERVAL_MS - elapsed);
      return;
    }
    const next = this._queue.shift();
    this._speak(next.key);
  }

  async _speak(key) {
    const text = this._resolveText(key);
    if (!text) { this._processQueue(); return; }

    this._speaking = true;
    this._lastSpoken[key] = Date.now();
    this._globalLast = Date.now();

    // Try Umuganda HTTP TTS for Kinyarwanda
    if (this._lang === 'rw') {
      const ok = await this._speakUmuganda(text);
      if (ok) return;
      // Falls through to Web Speech API with rw-RW language tag
    }

    this._speakWebSpeech(text);
  }

  _speakWebSpeech(text) {
    if (!this._synth) {
      this._speaking = false;
      this._processQueue();
      return;
    }
    const utt = new SpeechSynthesisUtterance(text);
    utt.lang  = this._lang === 'rw' ? 'rw-RW' : 'en-US';
    utt.rate  = 0.88;
    utt.pitch = 1.0;
    utt.volume = 1.0;

    utt.onend = () => {
      this._speaking = false;
      this._processQueue();
    };
    utt.onerror = () => {
      this._speaking = false;
      this._processQueue();
    };

    // Chrome bug: speechSynthesis hangs on long idle. Cancel first.
    this._synth.cancel();
    this._synth.speak(utt);
  }

  async _speakUmuganda(text) {
    try {
      const resp = await fetch(this._umugandaUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ text, speaker_id: 0 }),
        signal: AbortSignal.timeout(3000),
      });
      if (!resp.ok) return false;

      const buffer = await resp.arrayBuffer();
      if (!this._audioCtx) {
        this._audioCtx = new (window.AudioContext || window.webkitAudioContext)();
      }
      const audioBuffer = await this._audioCtx.decodeAudioData(buffer);
      const source = this._audioCtx.createBufferSource();
      source.buffer = audioBuffer;
      source.connect(this._audioCtx.destination);
      source.onended = () => {
        this._speaking = false;
        this._processQueue();
      };
      source.start(0);
      return true;
    } catch (_) {
      // Umuganda server not running — fall back to Web Speech API
      return false;
    }
  }

  _resolveText(key) {
    const map = this._lang === 'rw' ? PROMPTS_RW : PROMPTS_EN;
    return map[key] ?? PROMPTS_EN[key] ?? null;
  }
}
